airflow/models/trigger.py

```diff
@@ -201,10 +201,12 @@ def assign_unassigned(cls, triggerer_id, capacity, session=None):
             # notin_ doesn't find NULL rows
             .filter(or_(cls.triggerer_id.is_(None), cls.triggerer_id.notin_(alive_triggerer_ids)))
             .limit(capacity)
+            .with_for_update(skip_locked=True)
             .all()
         )
-        session.query(cls).filter(cls.id.in_([i.id for i in trigger_ids_query])).update(
-            {cls.triggerer_id: triggerer_id},
-            synchronize_session=False,
-        )
+        if trigger_ids_query:
+            session.query(cls).filter(cls.id.in_([i.id for i in trigger_ids_query])).update(
+                {cls.triggerer_id: triggerer_id},
+                synchronize_session=False,
+            )
```

Comment:
instead of doing a locking query and an update, why don't we just combine the query into the update statement?

then maybe we can avoid having to do any locking?  and we save a round trip to the database.  you think that would work?  so basically, just change the query to an update.

Reply:
I guess since we're trying to limit concurrency, we would need to do `update ... limit` and on postgres this is only achievable with subquery / CTE and ... at least judging by this [stackoverflow post](https://dba.stackexchange.com/a/69497), we would still need to do for update / skip locked....  although it would still probably be beneficial to avoid the round trip.

@andrewgodwin any concern for you with adding locking on the trigger table?

Reply:
I don't see why it would hurt, though I might suggest using locking columns rather than row-based locking if you want to add brand-new locking to something.

Reply:
@dstandish  update with subquery targeting same table is not supported in mysql.

Thats the reason I used locking.

Reply:
Maybe I can add locking only when multiple triggerer process is running, will that work?

Reply:
> 



> @dstandish update with subquery targeting same table is not supported in mysql.
> 
> Thats the reason I used locking.

cool makes sense

> Maybe I can add locking only when multiple triggerer process is running, will that work?

Probably not worth the trouble?

Reply:
Probably not. can we mark this discussion as resolved?